### DISCUSSION TITLE
Exclusive Fullscreen checker is broken on all renderers except legacy OpenGL + DirectX doesn't have exclusive fullscreen.

### DISCUSSION THREAD

**Discussion by Hasster1** [ppy/osu#22974] [Q&A]
Integrated Intel HD and Dedicated GeForce 840m (on which osu is being played), Windows 10 1909.

On all renderers, except legacy OGL, the game always says that it's in the exclusive fullscreen mode, even though it's clearly not (the game is below the DWM).
![broken](https://user-images.githubusercontent.com/20757715/227795611-e24f9659-0a07-44d4-87d6-6df278fa23f6.png)

Because i've got 2 GPUs on my laptop, and because osu is running on the dedicated one (not connected to the monitor), every app that doesn't have DirectX renderer (and even some that do have it) can't go into the exclusive fullscreen mode, no matter what i try.

The only way to enter exclusive fullscreen for me - is either to switch to an integrated GPU, or to disable fullscreen optimizations on an app, that works under DX renderer.

I really hoped that new Veldrid's DirectX would finally help me play osu at a higher fps and with an exclusive fullscreen mode, but it doesn't work at all, even when i boot it through my integrated GPU. Veldrid's OGL does have exclusive fullscreen.

**Comment by Hasster1** [comment]
Upd: There is no exclusive fullscreen even when i boot it through my integrated gpu. OGL does have it. I should've checked that before i wrote this whole thing.

**Comment by BlueZenithAm** [comment]
DX11 use filp mode to emulate the exclusive fullscreen performance. It should be as low latency as before since tearing is allowed in "exclusive fullscreen".

**Reply by Hasster1** [reply]
bweh. It's sad, but understandable. Unfortunately, for me flip mode presentation feels more like almost exclusive fullscreen with more severe screen tearing. Oh well, at least it's still better than borderless.

**Reply by BlueZenithAm** [reply]
> bweh. It's sad, but understandable. Unfortunately, for me flip mode presentation feels more like almost exclusive fullscreen with more severe screen tearing. Oh well, at least it's still better than borderless.

I wondering is it possible to let “allow tearing” become an option for DX renderer? I used to play with windowed mode and DWM off when I using Windows 7 (long time ago). Since I saw Windows 11 can enable “Optimization for windowed games”, but it doesn’t work in osu! with DX renderer.

**Reply by vjdato21** [reply]
While flip mode presentation gives an exclusive full-screen-like latency and easier `Alt` + `Tab`, it still has some downsides for me. I'm using a NVIDIA Optimus system and I can observe that when you try to stream using OBS/Discord, a huge amount of input latency is felt.

In comparison, osu! stable running compatibility mode doesn't feel like this when streaming using OBS.
 
Anyway, I don't know if this issue still persists in modern NVIDIA Optimus systems since new laptops have blazing fast RAMs now. My current laptop is from 2016.

**Comment by SupDos** [comment]
In my case (1 GPU system), only DirectX refuses to go into exclusive fullscreen, no matter how many "disable fullscreen optimizations" checkboxes I tick. OpenGL and Legacy handle it correctly. I think Vulkan also had the same issue, but that's been disabled for now

https://github.com/ppy/osu/discussions/22974#discussioncomment-5435437 mentioned that DX11 uses filp mode which is why it can't go into exclusive fulscreen, but from what I remember I thought that only DX12 has no support for exclusive fullscreen at all? So DX11 should still be able to support it correctly

**Reply by vjdato21** [reply]
Yes, DX11 should support it. But I guess it was done on purpose? I hope that they'll consider adding a tick-box for flip mode presentation in DX11 and other applicable APIs.

**Reply by SupDos** [reply]
Could also just disable flip mode entirely - if someone wants to play in borderless they will choose borderless, not fullscreen

**Reply by vjdato21** [reply]
> Could also just disable flip mode entirely - if someone wants to play in borderless they will choose borderless, not fullscreen

oh ye, and they can just turn on flip mode for borderless lel
